hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -14,8 +33,587 @@ import HIE.Bios as BIOS
+-- FilePath is needed for the special case that the Component is an Exe.
+-- The Exe contains a Path to the Main which is relative to some entry
+-- in 'ciSourceDirs'.
+-- We monkey-patch this by supplying the FilePath we want to load,
+-- which is part of this component, and select the 'ciSourceDir' we actually want.
+-- See the Documentation of 'ciSourceDir' to why this contains multiple entries.
+getTargets :: ChComponentInfo -> FilePath -> [String]
+getTargets comp fp = case ciEntrypoints comp of
+  ChSetupEntrypoint {} -> []
+  ChLibEntrypoint { chExposedModules, chOtherModules }
+    -> map unChModuleName (chExposedModules ++ chOtherModules)
+  ChExeEntrypoint { chMainIs, chOtherModules }
+    -> [sourceDir </> chMainIs | Just sourceDir <- [sourceDirs]]
+    ++ map unChModuleName chOtherModules
+    where
+      sourceDirs = find (`isFilePathPrefixOf` fp) (ciSourceDirs comp)
+
+-- | For all packages in a project, find the project the given FilePath
+-- belongs to most likely.
+findPackageFor :: NonEmpty (Package pt) -> FilePath -> Maybe (Package pt)
+findPackageFor packages fp = packages
+  & NonEmpty.toList
+  & sortOn (Down . pSourceDir)
+  & filter (\p -> pSourceDir p `isFilePathPrefixOf` fp)
+  & listToMaybe
+
+
+projectRootDir :: ProjLoc qt -> FilePath
+projectRootDir ProjLocV1CabalFile { plProjectDirV1 } = plProjectDirV1
+projectRootDir ProjLocV1Dir { plProjectDirV1 } = plProjectDirV1
+projectRootDir ProjLocV2File { plProjectDirV2 } = plProjectDirV2
+projectRootDir ProjLocV2Dir { plProjectDirV2 } = plProjectDirV2
+projectRootDir ProjLocStackYaml { plStackYaml } = takeDirectory plStackYaml
+
+projectSuffix :: ProjLoc qt -> FilePath
+projectSuffix ProjLocV1CabalFile {} = "Cabal-V1"
+projectSuffix ProjLocV1Dir {} = "Cabal-V1-Dir"
+projectSuffix ProjLocV2File {} = "Cabal-V2"
+projectSuffix ProjLocV2Dir {} = "Cabal-V2-Dir"
+projectSuffix ProjLocStackYaml {} = "Stack"
+
+-- | The hie-bios stack cradle doesn't return the target as well, so add the
+-- FilePath onto the end of the options to make sure at least one target
+-- is returned.
+fixCradle :: BIOS.Cradle -> BIOS.Cradle
+fixCradle cradle =
+  -- Normally this would also succeed for the 'Cabal-Helper-Stack' cradle.
+  -- Make sure that the cradle is definitely the one created by "HIE.Bios.Cradle.loadCradle"
+  if isStackCradle cradle
+  then
+    -- We need a lens
+    cradle { BIOS.cradleOptsProg =
+                (BIOS.cradleOptsProg
+                  cradle) { BIOS.runCradle = \fp' -> fmap (addOption fp')
+                              <$> BIOS.runCradle
+                                (BIOS.cradleOptsProg cradle)
+                                fp'
+                          }
+            }
+  else cradle
+  where
+    addOption fp (BIOS.ComponentOptions os ds) =
+      BIOS.ComponentOptions (os ++ [fp]) ds
+
+-- ----------------------------------------------------------------------------
+--
+-- Utility functions to manipulate FilePath's
+--
```

Comment:
I think it would be wise to build these up from the primitives in https://hackage.haskell.org/package/filepath, which take cognizance of platform differences, and are battle tested.

Reply:
I actually opened an issue upstream (https://github.com/haskell/filepath/issues/73) to add these functions to `filepath`, because I did not find a way to express them with the given API easily. 
They are based on `filepath`, though.